phoenix-core/src/main/java/org/apache/phoenix/schema/types/PDataType.java

```diff
@@ -1144,17 +1144,18 @@ public static PDataType fromLiteral(Object value) {
       return null;
     }
     for (PDataType type : PDataType.values()) {
-      if (type.isArrayType()) {
-        PhoenixArray arr = (PhoenixArray) value;
-        if ((type.getSqlType() == arr.baseType.sqlType + PDataType.ARRAY_TYPE_BASE)
-            && type.getJavaClass().isInstance(value)) {
-          return type;
-        }
-      } else {
-        if (type.getJavaClass().isInstance(value)) {
-          return type;
-        }
-      }
+			if (type.isArrayType() && type.getJavaClass().isInstance(value)) {
```

Comment:
Default implementation of isArrayType() returns false, hence I have not modified it. But in ordered list PVarcharArray comes before PJsonDataType and so value is first testified against PVarcharArraay. Now since PVarcharArray is ArrayType it gets inside the if() and type cast the value which throws ClassCastException since value is of type PJsonDataType. After casting the value, in next if (Line 1150) it validates whether this isInstance() which is not correct. Instead before casting value we should check whether it is instance of current type or not to avoid any ClassCastException.